crates/uniswap-v4/src/uniswap/pool_manager.rs

```diff
@@ -153,36 +163,37 @@ where
         let provider = Arc::clone(&self.provider);
         let filter = self.filter();
         let state_change_cache = Arc::clone(&self.state_change_cache);
+        let block_sync = self.block_sync.clone();
+
         let updated_pool_handle = tokio::spawn(async move {
-            let mut block_stream: BoxStream<Option<u64>> = provider.subscribe_blocks();
-            while let Some(block_number) = block_stream.next().await {
-                let chain_head_block_number =
-                    block_number.ok_or(PoolManagerError::EmptyBlockNumberFromStream)?;
+            let mut block_stream: BoxStream<Option<_>> = provider.subscribe_blocks();
+            while let Some(Some(block_number)) = block_stream.next().await {
                 // If there is a reorg, unwind state changes from last_synced block to the
                 // chain head block number
-                let mut reorg = false;
-                if chain_head_block_number <= last_synced_block {
-                    tracing::trace!(
-                        chain_head_block_number,
-                        last_synced_block,
-                        "reorg detected, unwinding state changes"
-                    );
-                    reorg = true;
-                    // set the last synced block to the head block number
-                    last_synced_block = chain_head_block_number - 1;
-                }
+
+                let (chain_head_block_number, block_range, is_reorg) = match block_number {
+                    PoolMangerBlocks::NewBlock(block) => (block, None, false),
+                    PoolMangerBlocks::Reorg(tip, range) => {
+                        last_synced_block = tip - range.end();
+                        tracing::trace!(
+                            tip,
+                            last_synced_block,
+                            "reorg detected, unwinding state changes"
+                        );
+                        (tip, Some(range), true)
+                    }
+                };
 
                 let logs = provider
                     .get_logs(
                         &filter
                             .clone()
-                            // last_synced_block + 1 == chain_head_block_number (always)
```

Comment:
Little frontrun here but this is incorrect as even though its super rare. you can still have more than 1 reorged blocks
![image](https://github.com/user-attachments/assets/d11b2c82-1a61-4b66-8bea-ec304ea98547)

Reply:
The length of the reorg doesn't matter we unwind from the last commited block and continue from there.

Reply:
Or at least that's how it's supposed to work anyway

Reply:
the previous unwind was always set to 1 though so in a mutliblock reorg, it would only handle the top block

Reply:
🤔 If we were at block 10, and then we get block 5, we unwind to 4 and apply 5. That's what this is supposed to be doing.

Reply:
![image](https://github.com/user-attachments/assets/0653a0d0-ca1b-42c9-b83c-a8a7ce347f58)
it always subtracting 1 would be the issue

Reply:
> 🤔 ? If we were at block 10, and then we get block 5, we unwind to 4 and apply 5 after the if. That's what this is supposed to be doing.

Thats not the way that reth cannonical updates work. They just give you the new chain.  in which you take the tip so even if there is a 5 block deep reorg. it will just be a new block update where in the update, the current tip will be the same block.
![image](https://github.com/user-attachments/assets/b805dc46-0805-4b0f-b5ec-a9704187e7f7)

Reply:
I don't get why that would prevent the code that's already there to work correctly but ok.

Reply:
because lets say were block 10. Reorg happens thats 5 blocks deep. last_synced_block = 10 and then we get a new event of the reorg. However in your impl there will be a single event that says. New block = 10. This will trigger the reorg where we will unwind to block 9 and then apply the last block in the reorg.

Reply:
Ah you are right.